### PR TITLE
Added TransferList Validation Example - RSC-1436

### PR DESCRIPTION
**Pull request by entroform**
https://issues.sonatype.org/browse/RSC-1436

### CODE REVIEW

File: gallery/src/styles/NxFormLayout/NxFormLayoutExample.tsx [hunk 1]
```diff
@@ -97,9 +97,19 @@ export default function NxFormLayoutExample() {
...
 
-  const [selectedTransferItems, setSelectedTransferItems] = useState<Set<number>>(new Set()),
+  const [transferItemsIsPristine, setTransferItemsIsPristine] = useState<boolean>(true),
+      [selectedTransferItems, setSelectedTransferItems] = useState<Set<number>>(new Set()),
       [availableTransferItemsFilter, setAvailableTransferItemsFilter] = useState(''),
-      [selectedTransferItemsFilter, setSelectedTransferItemsFilter] = useState('');
+      [selectedTransferItemsFilter, setSelectedTransferItemsFilter] = useState(''),
+      [transferItemsValidationErrors, setTransferItemsValidationErrors] = useState<string | null>(null);
+
+  function transferItemsOnChange(selectedItems: Set<number>) {
```
Comment (rpokorny) [inline gallery/src/styles/NxFormLayout/NxFormLayoutExample.tsx:106]: Would it make sense to create another hook to manage this, similar to `useCheckboxGroupState`?

Comment (entroform) [reply gallery/src/styles/NxFormLayout/NxFormLayoutExample.tsx:106]: > Would it make sense to create another hook to manage this, similar to `useCheckboxGroupState`?

Yup! I was about to ask about that in the meetings.

Comment (entroform) [reply gallery/src/styles/NxFormLayout/NxFormLayoutExample.tsx]: > > Would it make sense to create another hook to manage this, similar to `useCheckboxGroupState`?
> 
> Yup! I was about to ask about that in the meetings.

Added the hook here: 8ccfdf27dbf9f71a7f1add5c57a6057dc62b2747

File: gallery/src/components/NxFieldset/NxFieldsetPage.tsx [hunk 2]
```diff
@@ -331,6 +331,37 @@ const NxFieldsetPage = () =>
...
+              <NxTable.Cell>
+                <NxList className="nx-list--bulleted">
+                  <NxList.Item>
+                    <NxList.Text>
+                      <NxCode>initialValue</NxCode> initial <NxCode>NxTransferList</NxCode> selectedItems value
+                    </NxList.Text>
+                  </NxList.Item>
+                  <NxList.Item>
+                    <NxList.Text>
+                      <NxCode>validator</NxCode> function which will be run on each change in
```
Comment (EmilyBermudez88) [inline gallery/src/components/NxFieldset/NxFieldsetPage.tsx:345]:
```suggestion
                      <NxCode>validator</NxCode> function which will be run on each change in{' '}
```

**Comment by EmilyBermudez88** [comment]
Would it be worth adding an NxTransferList example into the Required/Validation Example on the NxFieldset page, since the `useTransferListState` is mentioned? Just a thought though

**Comment by entroform** [comment]
> Would it be worth adding an NxTransferList example into the Required/Validation Example on the NxFieldset page, since the `useTransferListState` is mentioned? Just a thought

Oo that's a good idea, thanks for catching that!

File: gallery/src/styles/NxFormLayout/NxFormLayoutExample.tsx (hunk 1 above)
Comment (rpokorny) [reply gallery/src/styles/NxFormLayout/NxFormLayoutExample.tsx]: > Added the hook here: https://github.com/sonatype/sonatype-react-shared-components/commit/8ccfdf27dbf9f71a7f1add5c57a6057dc62b2747

Actually I guess just like checkbox and radio fieldsets we should really start with pure-function state helpers (analogous to `checkboxGroupUserInput` etc) and then build the react hook on top of that.  We want to make things easily usable whether the validation is happening within a react component or somewhere else (like in redux)

Comment (entroform) [reply gallery/src/styles/NxFormLayout/NxFormLayoutExample.tsx]: > > Added the hook here: [8ccfdf2](https://github.com/sonatype/sonatype-react-shared-components/commit/8ccfdf27dbf9f71a7f1add5c57a6057dc62b2747)
> 
> Actually I guess just like checkbox and radio fieldsets we should really start with pure-function state helpers (analogous to `checkboxGroupUserInput` etc) and then build the react hook on top of that. We want to make things easily usable whether the validation is happening within a react component or somewhere else (like in redux)

Added that here: 7caa43015d039e3e6d9f7740338ffaf2656ad6cc

**Comment by entroform** [comment]
> Would it be worth adding an NxTransferList example into the Required/Validation Example on the NxFieldset page, since the `useTransferListState` is mentioned? Just a thought

Added the example here! d0f85eebdf5485e744888b268ed67deacad802ce

**Comment by entroform** [comment]
> Not something as a result of this ticket, but should probably be edited: in the Form Layout Page under `Component Summary`, NxFieldset is listed as a wrapper for checkboxes, radios and toggles, but no mention of transfer lists.

Fixed that here :)
7caa43015d039e3e6d9f7740338ffaf2656ad6cc

File: lib/src/components/NxFieldset/types.ts [hunk 3]
```diff
@@ -44,8 +45,20 @@ export interface CheckboxStateProps<T extends string = string> {
   validationErrors: ValidationErrors;
 }
 
+export interface TransferListProps<K> {
```
Comment (rpokorny) [inline lib/src/components/NxFieldset/types.ts:48]:
```suggestion
export interface TransferListStateProps<K> {
```

**Comment by rpokorny** [comment]
Should be added to the "Required/Validation Examples using Pure State Helpers" example

File: gallery/src/components/NxFieldset/NxFieldsetRequiredExample.tsx [hunk 4]
```diff
@@ -30,6 +39,19 @@ export default function NxFieldsetRequiredExample() {
...
   }, selectedDirs => selectedDirs.length ? null : requiredMessage);
 
+  const items = map<number, DataItem<number>>(i => ({ id: i, displayName: `Item ${i}` }), range(1, 101));
+  const {
+    state: [selectedItems, setSelectedItems],
+    isPristine: selectedItemsIsPristine,
+    validationErrors: selectedItemsValidationErrors
+  } = useTransferListState(
+      new Set<number>(),
+      (selectedItems) => selectedItems.size ? 'At least one item must be selected' : null
```
Comment (EmilyBermudez88) [inline gallery/src/components/NxFieldset/NxFieldsetRequiredExample.tsx:49]:
```suggestion
      (selectedItems) => !selectedItems.size ? 'At least one item must be selected' : null
```

**Comment by entroform** [comment]
> Should be added to the "Required/Validation Examples using Pure State Helpers" example

Added example here: 3da59d130d744a14b878d67c8c51e2ede60dff1b

File: gallery/src/components/NxFieldset/NxFieldsetPage.tsx [hunk 5]
```diff
@@ -331,6 +331,37 @@ const NxFieldsetPage = () =>
                 any validation errors based on those values, and whether the checkbox group is pristine.
               </NxTable.Cell>
             </NxTable.Row>
+            <NxTable.Row>
+              <NxTable.Cell><NxCode>useTransferListState</NxCode></NxTable.Cell>
```
Comment (rpokorny) [inline gallery/src/components/NxFieldset/NxFieldsetPage.tsx:335]: Pure state helpers need docs

Comment (entroform) [reply gallery/src/components/NxFieldset/NxFieldsetPage.tsx:389]: > Pure state helpers need docs

Ops, Added it here: dfe965cd1e72f3801f852125b396bf5863a71c3f

**Review by rpokorny — APPROVED** [review]

File: gallery/src/components/NxFieldset/NxFieldsetRequiredPureStateHelperExample.tsx [hunk 6]
```diff
@@ -5,48 +5,87 @@
...
+  allColors: DataItem<string>[];
+  availableColorsFilter: string;
+  selectedColorsFilter: string;
 }
 
 const requiredMessage = 'At least one color must be selected',
     radioValidator = (v: string | null) => v ? null : requiredMessage,
-    checkboxValidator = (selectedDirs: string[]) => selectedDirs.length ? null : requiredMessage;
+    checkboxValidator = (selectedDirs: string[]) => selectedDirs.length ? null : requiredMessage,
+    transferListValidator = (selectedItems: Set<string>) => selectedItems ? null : requiredMessage;
```
Comment (EmilyBermudez88) [inline gallery/src/components/NxFieldset/NxFieldsetRequiredPureStateHelperExample.tsx:38]:
```suggestion
    transferListValidator = (selectedItems: Set<string>) => selectedItems.size ? null : requiredMessage;
```

**Review by EmilyBermudez88 — APPROVED** [review]

**Pull request merged**